File: zk/prover/hedgeCanonical.ts
/**
 * Canonical serialization + hashing of private-hedge inputs — TypeScript side.
 *
 * MUST produce byte-identical output to `zkp/core/hedge_canonical.py` so
 * the browser prover client (soon) and Python STARK server agree on the
 * commitment hash, and Move on-chain verifier consumes it as-is.
 *
 * Purpose: a depositor commits to a hedge (asset, side, size, leverage,
 * entry price, salt) with a single 32-byte hash on-chain. The proof
 * proves off-chain that the hidden inputs satisfy vault invariants; the
 * `public_inputs` pin the CAPS the proof was made against so any
 * observer can see the safety envelope even without the trade details.
 *
 * Precision conventions (must match Python):
 *   - USDC-denominated values → integer cents ($1 = 100). Wire encoded
 *     as u128-BE (BigInt in TS) so trillion-dollar caps are safe.
 *   - Leverage                → integer x (1..leverageCap). u32.
 *   - Prices                  → integer USDC cents. u64.
 *   - Sizes                   → integer per-asset step units. u64.
 *   - Asset / side            → uppercase ASCII, enum-coded on wire.
 *   - Salt                    → 32 raw bytes (64 hex chars, lowercase).
 *   - Timestamps              → ms epoch, floored to nearest 1000 ms.
 */

import crypto from 'crypto';

/** Sentinel — MUST match HEDGE_CANONICAL_VERSION in zkp/core/hedge_canonical.py. */
export const HEDGE_CANONICAL_VERSION = 1 as const;

/**
 * Enum codes go into the fixed-layout commitment_hash preimage. NEVER
 * renumber — existing on-chain commitments would drift.
 */
export const ASSET_CODE = {
  BTC: 1,
  ETH: 2,
  SUI: 3,
} as const;

export const SIDE_CODE = {
  LONG: 0,
  SHORT: 1,
} as const;

export type HedgeAsset = keyof typeof ASSET_CODE;
export type HedgeSide = keyof typeof SIDE_CODE;

export const ALLOWED_ASSETS: ReadonlySet<HedgeAsset> = new Set(
  Object.keys(ASSET_CODE) as HedgeAsset[],
);
export const ALLOWED_SIDES: ReadonlySet<HedgeSide> = new Set(
  Object.keys(SIDE_CODE) as HedgeSide[],
);

/** Thrown when a hedge witness fails a binding invariant (client-side pre-flight). */
export class HedgeBindingError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'HedgeBindingError';
  }
}

/**
 * The exact schema the STARK proof binds. Fields are integers or
 * bigint-string integers to keep serialization language-neutral.
 *
 * `sizeUnits`, `notionalValueUsdcCents`, `notionalCapUsdcCents` are
 * `bigint` so cent-precision doesn't overflow 2^53. Other fields fit
 * in Number.
 */
export interface CanonicalHedgeInputs {
  version: typeof HEDGE_CANONICAL_VERSION;
  chain: string;
  portfolioId: number;
  /** ms epoch — floored to nearest 1000 ms so proof and attestation share the same second. */
  timestampMs: number;
  asset: HedgeAsset;
  side: HedgeSide;
  /** Position size in per-asset step units (SUI: 1 = 1 SUI, BTC: 1000 = 0.001 BTC, ETH: 10000 = 0.01 ETH). */
  sizeUnits: bigint;
  leverageX: number;
  /** Entry price in USDC cents (u64). */
  entryPriceUsdcCents: bigint;
  /** size × entry × 1× (base notional). u128 on wire. */
  notionalValueUsdcCents: bigint;
  /** Max allowed leverage for this hedge — pinned in commitment. */
  leverageCap: number;
  /** Max allowed notional for this hedge — pinned in commitment. */
  notionalCapUsdcCents: bigint;
  /** 32-byte hex, lowercase, no 0x prefix. */
  salt: string;
}

function normalize(inputs: CanonicalHedgeInputs): CanonicalHedgeInputs {
  return {
    version: HEDGE_CANONICAL_VERSION,
    chain: inputs.chain.toLowerCase(),
    portfolioId: Math.trunc(inputs.portfolioId),
    timestampMs: Math.floor(inputs.timestampMs / 1000) * 1000,
    asset: inputs.asset.toString().toUpperCase() as HedgeAsset,
    side: inputs.side.toString().toUpperCase() as HedgeSide,
    sizeUnits: BigInt(inputs.sizeUnits),
    leverageX: Math.trunc(inputs.leverageX),
    entryPriceUsdcCents: BigInt(inputs.entryPriceUsdcCents),
    notionalValueUsdcCents: BigInt(inputs.notionalValueUsdcCents),
    leverageCap: Math.trunc(inputs.leverageCap),
    notionalCapUsdcCents: BigInt(inputs.notionalCapUsdcCents),
    salt: inputs.salt.toLowerCase(),
  };
}

/**
 * Canonical JSON. Sorted keys at every level, no whitespace. `bigint`
 * fields serialize as bare integer literals (no quotes) so Python's
 * `json.loads`-parsed output byte-matches ours.
 */
export function serializeCanonical(inputs: CanonicalHedgeInputs): string {
  return stringifySortedKeys(normalize(inputs));
}

function stringifySortedKeys(v: unknown): string {
  if (v === null) return 'null';
  if (typeof v === 'bigint') return v.toString();
  if (typeof v === 'number') {
    if (!Number.isFinite(v)) throw new HedgeBindingError(`non-finite number: ${v}`);
    return v.toString();
  }
  if (typeof v === 'boolean') return v ? 'true' : 'false';
  if (typeof v === 'string') return JSON.stringify(v);
  if (Array.isArray(v)) return `[${v.map(stringifySortedKeys).join(',')}]`;
  if (typeof v === 'object') {
    const obj = v as Record<string, unknown>;
    const keys = Object.keys(obj).sort();
    return `{${keys.map((k) => `${JSON.stringify(k)}:${stringifySortedKeys(obj[k])}`).join(',')}}`;
  }
  throw new HedgeBindingError(`unserializable value: ${typeof v}`);
}

/** SHA-256 of the canonical JSON serialization. Byte-matches Python. */
export function computeInputsHash(inputs: CanonicalHedgeInputs): string {
  const bytes = Buffer.from(serializeCanonical(inputs), 'utf8');
  return crypto.createHash('sha256').update(bytes).digest('hex');
}

function u32BE(n: number): Buffer {
  const buf = Buffer.alloc(4);
  buf.writeUInt32BE(n >>> 0, 0);
  return buf;
}

function u64BE(n: bigint): Buffer {
  const buf = Buffer.alloc(8);
  buf.writeBigUInt64BE(BigInt.asUintN(64, n));
  return buf;
}

function u128BE(n: bigint): Buffer {
  if (n < 0n || n >= 1n << 128n) throw new HedgeBindingError(`u128 out of range: ${n}`);
  const buf = Buffer.alloc(16);
  const hi = n >> 64n;
  const lo = n & 0xffffffffffffffffn;
  buf.writeBigUInt64BE(BigInt.asUintN(64, hi), 0);
  buf.writeBigUInt64BE(BigInt.asUintN(64, lo), 8);
  return buf;
}

function u8(n: number): Buffer {
  return Buffer.from([n & 0xff]);
}

/**
 * 32-byte SHA256 the depositor's key signs and Move verifies on-chain.
 * Fixed binary layout — NEVER change without a version bump.
 *
 *   SHA256(
 *     version_u32BE                  (4)
 *     portfolioId_u32BE              (4)
 *     timestampMs_u64BE              (8)
 *     asset_code_u8                  (1)
 *     side_code_u8                   (1)
 *     leverageX_u32BE                (4)
 *     leverageCap_u32BE              (4)
 *     entryPriceUsdcCents_u64BE      (8)
 *     sizeUnits_u64BE                (8)
 *     notionalValueUsdcCents_u128BE  (16)
 *     notionalCapUsdcCents_u128BE    (16)
 *     salt_32B                       (32)
 *     inputsHash_32B                 (32)
 *   )
 *   = 146 bytes.
 */
export function computeCommitmentHash(
  inputs: CanonicalHedgeInputs,
  inputsHash: string,
): string {
  const n = normalize(inputs);

  const assetCode = (ASSET_CODE as Record<string, number>)[n.asset];
  if (assetCode == null) throw new HedgeBindingError(`unsupported asset: ${n.asset}`);
  const sideCode = (SIDE_CODE as Record<string, number>)[n.side];
  if (sideCode == null) throw new HedgeBindingError(`unsupported side: ${n.side}`);

  const saltBytes = Buffer.from(n.salt, 'hex');
  if (saltBytes.length !== 32) {
    throw new HedgeBindingError(
      `salt must be 32 bytes (64 hex chars), got ${saltBytes.length} bytes`,
    );
  }
  const inputsHashBytes = Buffer.from(inputsHash, 'hex');
  if (inputsHashBytes.length !== 32) {
    throw new HedgeBindingError(
      `inputs_hash must be 32 bytes (64 hex chars), got ${inputsHashBytes.length} bytes`,
    );
  }

  const preimage = Buffer.concat([
    u32BE(HEDGE_CANONICAL_VERSION),
    u32BE(n.portfolioId),
    u64BE(BigInt(n.timestampMs)),
    u8(assetCode),
    u8(sideCode),
    u32BE(n.leverageX),
    u32BE(n.leverageCap),
    u64BE(n.entryPriceUsdcCents),
    u64BE(n.sizeUnits),
    u128BE(n.notionalValueUsdcCents),
    u128BE(n.notionalCapUsdcCents),
    saltBytes,
    inputsHashBytes,
  ]);
  return crypto.createHash('sha256').update(preimage).digest('hex');
}

/**
 * Client-side pre-flight of the same invariants the Python server
 * enforces via `assert_hedge_binding`. Callers should run this before
 * submitting to `/api/zk/attest` — surfaces the error locally instead
 * of via a 400 response.
 */
export function assertHedgeBindingLocal(
  statement: { claim: string; public_inputs: Array<string | number | bigint> },
  witness: { canonical: CanonicalHedgeInputs },
): CanonicalHedgeInputs {
  const expectedClaim = `zkv-hedge-v${HEDGE_CANONICAL_VERSION}`;
  if (statement.claim !== expectedClaim) {
    throw new HedgeBindingError(
      `hedge-binding: claim mismatch — expected '${expectedClaim}', got '${statement.claim}'`,
    );
  }

  const pub = statement.public_inputs;
  if (!Array.isArray(pub) || pub.length < 4) {
    throw new HedgeBindingError(
      'hedge-binding: public_inputs must be [commitment_hash, inputs_hash, notional_cap_cents, leverage_cap]',
    );
  }

  const stmtCommitment = String(pub[0]).toLowerCase();
  const stmtInputsHash = String(pub[1]).toLowerCase();
  const stmtNotionalCap = BigInt(pub[2]);
  const stmtLeverageCap = Number(pub[3]);

  const canonical = witness.canonical;
  if (!canonical || typeof canonical !== 'object') {
    throw new HedgeBindingError('hedge-binding: witness.canonical is required');
  }

  const asset = String(canonical.asset).toUpperCase() as HedgeAsset;
  if (!ALLOWED_ASSETS.has(asset)) {
    throw new HedgeBindingError(
      `hedge-binding: asset '${asset}' not in allow-list ${[...ALLOWED_ASSETS].sort().join(',')}`,
    );
  }
  const side = String(canonical.side).toUpperCase() as HedgeSide;
  if (!ALLOWED_SIDES.has(side)) {
    throw new HedgeBindingError(
      `hedge-binding: side '${side}' not in allow-list ${[...ALLOWED_SIDES].sort().join(',')}`,
    );
  }

  const n = normalize(canonical);

  if (n.leverageCap !== stmtLeverageCap) {
    throw new HedgeBindingError(
      `hedge-binding: witness.leverageCap (${n.leverageCap}) != statement.leverage_cap (${stmtLeverageCap})`,
    );
  }
  if (n.notionalCapUsdcCents !== stmtNotionalCap) {
    throw new HedgeBindingError(
      `hedge-binding: witness.notionalCapUsdcCents (${n.notionalCapUsdcCents}) != statement.notional_cap_cents (${stmtNotionalCap})`,
    );
  }

  if (!(n.leverageX >= 1 && n.leverageX <= n.leverageCap)) {
    throw new HedgeBindingError(
      `hedge-binding: leverageX (${n.leverageX}) outside [1, ${n.leverageCap}]`,
    );
  }
  if (n.notionalValueUsdcCents <= 0n) {
    throw new HedgeBindingError(
      `hedge-binding: notionalValueUsdcCents must be positive, got ${n.notionalValueUsdcCents}`,
    );
  }
  if (n.notionalValueUsdcCents > n.notionalCapUsdcCents) {
    throw new HedgeBindingError(
      `hedge-binding: notional (${n.notionalValueUsdcCents}) exceeds cap (${n.notionalCapUsdcCents})`,
    );
  }
  if (n.sizeUnits <= 0n) {
    throw new HedgeBindingError(
      `hedge-binding: sizeUnits must be positive, got ${n.sizeUnits}`,
    );
  }
  if (n.entryPriceUsdcCents <= 0n) {
    throw new HedgeBindingError(
      `hedge-binding: entryPriceUsdcCents must be positive, got ${n.entryPriceUsdcCents}`,
    );
  }

  if (n.salt.length !== 64 || !/^[0-9a-f]{64}$/.test(n.salt)) {
    throw new HedgeBindingError(
      `hedge-binding: salt must be 64 lowercase hex chars, got '${n.salt}'`,
    );
  }

  const recomputedInputsHash = computeInputsHash(canonical);
  if (recomputedInputsHash !== stmtInputsHash) {
    throw new HedgeBindingError(
      `hedge-binding: inputs_hash mismatch — recomputed=${recomputedInputsHash} claimed=${stmtInputsHash}`,
    );
  }

  const recomputedCommitment = computeCommitmentHash(canonical, recomputedInputsHash);
  if (recomputedCommitment !== stmtCommitment) {
    throw new HedgeBindingError(
      `hedge-binding: commitment_hash mismatch — recomputed=${recomputedCommitment} claimed=${stmtCommitment}`,
    );
  }

  return n;
}

/**
 * Everything a caller needs to submit a bound STARK request. One-shot
 * canonicalize → hash → return.
 */
export interface HedgeBinding {
  canonical: CanonicalHedgeInputs;
  canonicalBytes: string;
  inputsHash: string;
  commitmentHash: string;
}

export function prepareHedgeBinding(inputs: CanonicalHedgeInputs): HedgeBinding {
  const canonicalBytes = serializeCanonical(inputs);
  const inputsHash = crypto.createHash('sha256').update(canonicalBytes, 'utf8').digest('hex');
  const commitmentHash = computeCommitmentHash(inputs, inputsHash);
  return { canonical: inputs, canonicalBytes, inputsHash, commitmentHash };
}

File: zkp/core/hedge_canonical.py
"""
Canonical serialization + hashing of private-hedge inputs — Python side.

MUST produce byte-identical output to `lib/hedge-canonical.ts` so the TS
prover client and Python STARK prover agree on the commitment hash the
depositor's key signs and the on-chain Move verifier consumes.

Purpose: a depositor can commit to a hedge (asset, side, size, leverage,
entry price, salt) with a single 32-byte hash on-chain, and prove
off-chain that the hidden inputs satisfy vault invariants (leverage
capped, notional capped, asset in allow-list, side well-formed). The
`public_inputs` pin the CAPS the proof was made against, so any observer
can see the safety envelope even without seeing the trade.

Precision conventions (must match TS):
  - USDC-denominated values → integer cents ($1 = 100). Big enough for
    trillion-dollar notional at u128, so u128-BE on the wire.
  - Leverage                → integer x (1..leverageCap). u32.
  - Prices                  → integer USDC cents. u64.
  - Sizes                   → integer per-asset step units (e.g. SUI:1,
    ETH: 0.01 → 10000 micro-units, BTC: 0.001 → 1000 milli-units). u64.
  - Asset symbols           → uppercase ASCII, enum-coded on the wire.
  - Side                    → "LONG" | "SHORT", enum-coded on the wire.
  - Salt                    → 32 raw bytes (64 hex chars, no 0x prefix).
  - Timestamps              → ms epoch, floored to nearest 1000 ms.
"""

import hashlib
import json
import struct
from typing import Any, Dict


HEDGE_CANONICAL_VERSION = 1

# Enum codes go into the fixed-layout commitment_hash preimage. Never
# renumber — existing on-chain commitments would drift.
ASSET_CODE = {
    "BTC": 1,
    "ETH": 2,
    "SUI": 3,
}

SIDE_CODE = {
    "LONG": 0,
    "SHORT": 1,
}

ALLOWED_ASSETS = frozenset(ASSET_CODE.keys())
ALLOWED_SIDES = frozenset(SIDE_CODE.keys())


class HedgeBindingError(ValueError):
    """
    Raised when a hedge proof request fails a binding check — canonical
    hash mismatch, cap violation, malformed side/asset, salt-length wrong,
    or claimed commitment_hash doesn't match the recomputation. Honest
    prover must not proceed past these checks.
    """


def _normalize(inputs: Dict[str, Any]) -> Dict[str, Any]:
    """Uppercase asset/side, floor timestamp to seconds, coerce numerics."""
    asset = str(inputs["asset"]).upper()
    side = str(inputs["side"]).upper()
    return {
        "version": HEDGE_CANONICAL_VERSION,
        "chain": str(inputs["chain"]).lower(),
        "portfolioId": int(inputs["portfolioId"]),
        "timestampMs": int(inputs["timestampMs"]) // 1000 * 1000,
        "asset": asset,
        "side": side,
        "sizeUnits": int(inputs["sizeUnits"]),
        "leverageX": int(inputs["leverageX"]),
        "entryPriceUsdcCents": int(inputs["entryPriceUsdcCents"]),
        "notionalValueUsdcCents": int(inputs["notionalValueUsdcCents"]),
        "leverageCap": int(inputs["leverageCap"]),
        "notionalCapUsdcCents": int(inputs["notionalCapUsdcCents"]),
        "salt": str(inputs["salt"]).lower(),
    }


def serialize_canonical(inputs: Dict[str, Any]) -> str:
    """
    JSON encode with sorted keys and compact separators — byte-identical
    to the TS `serializeCanonical`. This is the string SHA256 hashes for
    `inputs_hash`.
    """
    normalized = _normalize(inputs)
    return json.dumps(normalized, sort_keys=True, separators=(",", ":"))


def compute_inputs_hash(inputs: Dict[str, Any]) -> str:
    """SHA256 hex of the canonical JSON serialization."""
    return hashlib.sha256(serialize_canonical(inputs).encode("utf-8")).hexdigest()


def _u128_be(x: int) -> bytes:
    """Big-endian unsigned 128-bit encoding — no struct format supports u128 natively."""
    if x < 0 or x >= (1 << 128):
        raise HedgeBindingError(f"u128 out of range: {x}")
    return x.to_bytes(16, byteorder="big", signed=False)


def compute_commitment_hash(inputs: Dict[str, Any], inputs_hash_hex: str) -> str:
    """
    32-byte SHA256 the depositor's key signs and Move verifies on-chain.
    Fixed binary layout — NEVER change without a version bump (would
    break every existing commitment).

      SHA256(
        version_u32BE                  (4)
        portfolioId_u32BE              (4)
        timestampMs_u64BE              (8)
        asset_code_u8                  (1)   # from ASSET_CODE
        side_code_u8                   (1)   # from SIDE_CODE
        leverageX_u32BE                (4)
        leverageCap_u32BE              (4)
        entryPriceUsdcCents_u64BE      (8)
        sizeUnits_u64BE                (8)
        notionalValueUsdcCents_u128BE  (16)
        notionalCapUsdcCents_u128BE    (16)
        salt_32B                       (32)
        inputsHash_32B                 (32)
      )
      = 146 bytes total.
    """
    n = _normalize(inputs)

    asset_code = ASSET_CODE.get(n["asset"])
    if asset_code is None:
        raise HedgeBindingError(f"unsupported asset: {n['asset']!r}")
    side_code = SIDE_CODE.get(n["side"])
    if side_code is None:
        raise HedgeBindingError(f"unsupported side: {n['side']!r}")

    salt_bytes = bytes.fromhex(n["salt"])
    if len(salt_bytes) != 32:
        raise HedgeBindingError(
            f"salt must be 32 bytes (64 hex chars), got {len(salt_bytes)} bytes"
        )
    inputs_hash_bytes = bytes.fromhex(inputs_hash_hex)
    if len(inputs_hash_bytes) != 32:
        raise HedgeBindingError(
            f"inputs_hash must be 32 bytes (64 hex chars), got {len(inputs_hash_bytes)} bytes"
        )

    parts = b"".join(
        [
            struct.pack(">I", HEDGE_CANONICAL_VERSION),
            struct.pack(">I", n["portfolioId"] & 0xFFFFFFFF),
            struct.pack(">Q", n["timestampMs"] & 0xFFFFFFFFFFFFFFFF),
            struct.pack(">B", asset_code),
            struct.pack(">B", side_code),
            struct.pack(">I", n["leverageX"] & 0xFFFFFFFF),
            struct.pack(">I", n["leverageCap"] & 0xFFFFFFFF),
            struct.pack(">Q", n["entryPriceUsdcCents"] & 0xFFFFFFFFFFFFFFFF),
            struct.pack(">Q", n["sizeUnits"] & 0xFFFFFFFFFFFFFFFF),
            _u128_be(n["notionalValueUsdcCents"]),
            _u128_be(n["notionalCapUsdcCents"]),
            salt_bytes,
            inputs_hash_bytes,
        ]
    )
    return hashlib.sha256(parts).hexdigest()


def assert_hedge_binding(
    statement: Dict[str, Any], witness: Dict[str, Any]
) -> Dict[str, Any]:
    """
    Full binding check for a hedge proof. Called by the FastAPI server
    before invoking the STARK prover.

    Requires:
      statement.claim         == "zkv-hedge-v1"
      statement.public_inputs == [commitment_hash, inputs_hash,
                                  notional_cap_cents, leverage_cap]
      witness.canonical       == CanonicalHedgeInputs dict

    Enforces vault invariants: asset in allow-list, side well-formed,
    leverage in [1, leverageCap], notional ≤ cap, size × entry ≈ notional
    (integer-cent tolerance), salt is 32 bytes.

    Raises HedgeBindingError on any mismatch. Returns the normalized
    canonical dict for downstream logging / attestation.
    """
    claim = str(statement.get("claim") or "")
    expected_claim = f"zkv-hedge-v{HEDGE_CANONICAL_VERSION}"
    if claim != expected_claim:
        raise HedgeBindingError(
            f"hedge-binding: claim mismatch — expected '{expected_claim}', got '{claim}'"
        )

    public_inputs = statement.get("public_inputs") or []
    if not isinstance(public_inputs, list) or len(public_inputs) < 4:
        raise HedgeBindingError(
            "hedge-binding: public_inputs must be "
            "[commitment_hash, inputs_hash, notional_cap_cents, leverage_cap]"
        )

    stmt_commitment_hash = str(public_inputs[0]).lower()
    stmt_inputs_hash = str(public_inputs[1]).lower()
    try:
        stmt_notional_cap = int(public_inputs[2])
        stmt_leverage_cap = int(public_inputs[3])
    except (TypeError, ValueError) as e:
        raise HedgeBindingError(
            f"hedge-binding: notional_cap / leverage_cap must be integer-castable — {e}"
        )

    canonical = witness.get("canonical")
    if not isinstance(canonical, dict):
        raise HedgeBindingError(
            "hedge-binding: witness.canonical is required and must be a dict"
        )

    # 1) Whitelist enforcement (fail fast before any hashing).
    asset = str(canonical.get("asset", "")).upper()
    if asset not in ALLOWED_ASSETS:
        raise HedgeBindingError(
            f"hedge-binding: asset {asset!r} not in allow-list {sorted(ALLOWED_ASSETS)}"
        )
    side = str(canonical.get("side", "")).upper()
    if side not in ALLOWED_SIDES:
        raise HedgeBindingError(
            f"hedge-binding: side {side!r} not in allow-list {sorted(ALLOWED_SIDES)}"
        )

    normalized = _normalize(canonical)

    # 2) Cap pinning: caps advertised in public_inputs must match the
    # caps the witness was normalized against. Otherwise an attacker could
    # ship a witness with leverageCap=4 but public_inputs claiming
    # leverage_cap=2 — the proof would attest to safety it doesn't provide.
    if normalized["leverageCap"] != stmt_leverage_cap:
        raise HedgeBindingError(
            f"hedge-binding: witness.leverageCap ({normalized['leverageCap']}) "
            f"!= statement.leverage_cap ({stmt_leverage_cap})"
        )
    if normalized["notionalCapUsdcCents"] != stmt_notional_cap:
        raise HedgeBindingError(
            f"hedge-binding: witness.notionalCapUsdcCents ({normalized['notionalCapUsdcCents']}) "
            f"!= statement.notional_cap_cents ({stmt_notional_cap})"
        )

    # 3) Vault invariants on the actual hedge values.
    if not (1 <= normalized["leverageX"] <= normalized["leverageCap"]):
        raise HedgeBindingError(
            f"hedge-binding: leverageX ({normalized['leverageX']}) outside [1, "
            f"{normalized['leverageCap']}]"
        )
    if normalized["notionalValueUsdcCents"] <= 0:
        raise HedgeBindingError(
            f"hedge-binding: notionalValueUsdcCents must be positive, got "
            f"{normalized['notionalValueUsdcCents']}"
        )
    if normalized["notionalValueUsdcCents"] > normalized["notionalCapUsdcCents"]:
        raise HedgeBindingError(
            f"hedge-binding: notional ({normalized['notionalValueUsdcCents']}) exceeds cap "
            f"({normalized['notionalCapUsdcCents']})"
        )
    if normalized["sizeUnits"] <= 0:
        raise HedgeBindingError(
            f"hedge-binding: sizeUnits must be positive, got {normalized['sizeUnits']}"
        )
    if normalized["entryPriceUsdcCents"] <= 0:
        raise HedgeBindingError(
            f"hedge-binding: entryPriceUsdcCents must be positive, got "
            f"{normalized['entryPriceUsdcCents']}"
        )

    # 4) Salt shape (32 bytes = 64 hex chars, lowercase).
    salt = normalized["salt"]
    if len(salt) != 64 or any(c not in "0123456789abcdef" for c in salt):
        raise HedgeBindingError(
            f"hedge-binding: salt must be 64 lowercase hex chars, got {salt!r}"
        )

    # 5) Recompute inputs_hash and compare to statement.
    recomputed_inputs_hash = compute_inputs_hash(canonical)
    if recomputed_inputs_hash != stmt_inputs_hash:
        raise HedgeBindingError(
            "hedge-binding: inputs_hash mismatch — the witness does NOT hash to "
            f"the claimed inputsHash. recomputed={recomputed_inputs_hash} "
            f"claimed={stmt_inputs_hash}"
        )

    # 6) Recompute commitment_hash and compare to statement.
    recomputed_commitment = compute_commitment_hash(canonical, recomputed_inputs_hash)
    if recomputed_commitment != stmt_commitment_hash:
        raise HedgeBindingError(
            "hedge-binding: commitment_hash mismatch — recomputed commitment does not "
            f"match statement. recomputed={recomputed_commitment} "
            f"claimed={stmt_commitment_hash}"
        )

    return normalized
